src/tsolvers/egraph/Explainer.cc

```diff
@@ -0,0 +1,458 @@
+{
+    assert(getEnode(x).isTerm());
+    assert(getEnode(y).isTerm());
+    assert(getEnode(x).getRoot() != getEnode(y).getRoot());
+
+    // They must be different because the merge hasn't occured yet
+    if (getEnode(x).getRoot() == getEnode(y).getRoot()) {
+        throw OsmtInternalException(); // "Attempting to store explanation for already known equality"
+    }
+    // The main observation here is that the explanation tree, altough
+    // differently oriented, has the same size as the equivalence tree
+    // (actually we don't keep the equivalence tree, because we use
+    // the quick-find approach, but here we just need the size). So we
+    // can use root_x.getSize() to know the size of the class of a node
+    // x and therefore be sure that the explanation tree is kept
+    // balanced (which is a requirement to keep the O(nlogn) bound
+
+    // Make sure that x is the node with the larger number of edges to switch
+    const Enode& root_x = getEnode(getEnode(x).getRoot());
+    const Enode& root_y = getEnode(getEnode(y).getRoot());
+    if ( root_x.getSize() < root_y.getSize() ) {
+        ERef tmp = x;
+        x = y;
+        y = tmp;
+    }
+    // Reroot the explanation tree on y. It has an amortized cost of logn
+    reRootOn( y );
+
+    getEnode(y).setExpParent(x);
+    getEnode(y).setExpReason(reason);
+    // Store both nodes. Because of rerooting operations
+    // we don't know whether x --> y or x <-- y at the moment of
+    // backtracking. So we just save reason and check both parts
+    exp_undo_stack.push(x);
+    exp_undo_stack.push(y);
+
+#ifdef VERBOSE_EUF
+    assert( checkExpTree(getEnode(x).getTerm()) );
+    assert( checkExpTree(getEnode(y).getTerm()) );
+//    cout << printExplanationTree( tr_y ) << endl;
+#endif
+}
+
+//
+// Subroutine of explainStoreExplanation
+// Re-root the tree containing x, in such a way that
+// the new root is x itself
+//
+void Explainer::reRootOn(ERef x) {
+    ERef p = x;
+    ERef parent = getEnode(p).getExpParent();
+    PtAsgn reason = getEnode(p).getExpReason();
+    getEnode(x).setExpParent(ERef_Undef);
+    getEnode(x).setExpReason(PtAsgn_Undef);
+    while( parent != ERef_Undef ) {
+        // Save grandparent
+        ERef grandparent = getEnode(parent).getExpParent();
+        // Save reason
+        PtAsgn saved_reason = reason;
+        reason = getEnode(parent).getExpReason();
+        // Reverse edge & reason
+        getEnode(parent).setExpParent(p);
+        getEnode(parent).setExpReason(saved_reason);
+
+#ifdef VERBOSE_EUF
+        assert( checkExpTree( getEnode(parent).getTerm() ) );
+#endif
+
+        // Move the two pointers
+        p = parent;
+        parent = grandparent;
+    }
+}
+
+vec<PtAsgn> Explainer::explain(std::pair<ERef,ERef> nodePair) {
+
+#ifdef EXPLICIT_CONGRUENCE_EXPLANATIONS
+    congruences.clear();
+#endif
+
+    DupChecker dupChecker(dcd);
+    vec<PtAsgn> explanation;
+    PendingQueue exp_pending;
+    exp_pending.push(nodePair);
```

Comment:
What is the motivation behind passing these data structures (dupChecker, explanation, exp_pending) as arguments to subprocedures, instead of having them as members (which I believe was the original design)?

I am conflicted about this as well, I see both advantages and disadvantages.

Reply:
The conclusion is to keep this version.